File: src/features/typing/data/java/number_of_islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int total = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    total++;
                    sink(grid, rows, cols, i, j);
                }
            }
        }
        return total;
    }

    private void sink(char[][] grid, int rows, int cols, int i, int j) {
        if (i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] != '1') {
            return;
        }
        grid[i][j] = '0';
        sink(grid, rows, cols, i + 1, j);
        sink(grid, rows, cols, i - 1, j);
        sink(grid, rows, cols, i, j + 1);
        sink(grid, rows, cols, i, j - 1);
    }
}

File: src/features/typing/data/dart/number_of_islands.dart
class Solution {
  int numIslands(List<List<String>> grid) {
    final rows = grid.length;
    final cols = grid[0].length;
    var total = 0;
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (grid[i][j] == '1') {
          total++;
          _sink(grid, rows, cols, i, j);
        }
      }
    }
    return total;
  }

  void _sink(List<List<String>> grid, int rows, int cols, int i, int j) {
    if (i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] != '1') {
      return;
    }
    grid[i][j] = '0';
    _sink(grid, rows, cols, i + 1, j);
    _sink(grid, rows, cols, i - 1, j);
    _sink(grid, rows, cols, i, j + 1);
    _sink(grid, rows, cols, i, j - 1);
  }
}
